### PR TITLE
Improve hatch demo

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Minor improvement to the hatch demo:

- more cross ref.
- OSX actually has hatches.  Not sure about WX or Cairo.  Updated that in `hatch_style_reference` and added a couple of crosslinks.

New demo is at:

https://41352-1385122-gh.circle-artifacts.com/0/doc/build/html/gallery/shapes_and_collections/hatch_demo.html#sphx-glr-gallery-shapes-and-collections-hatch-demo-py

Lightly modified hatch_reference_style.py: 

https://41352-1385122-gh.circle-artifacts.com/0/doc/build/html/gallery/shapes_and_collections/hatch_style_reference.html#sphx-glr-gallery-shapes-and-collections-hatch-style-reference-py

Figure_1

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
As noted before, I would combine all 4 hatch demos, but this at least improves the existing demo...

**Comment by jklymak** [comment]
> This removes important recently added information (density and combination) from the reference. And it instead adds back in the ugly plot from the hatch demo. Is that intentional?

I made a mistake with the reference file.  It was supposed to just change the first paragraph, and that's been fixed.

Not sure what the second half of your sentence means.  This PR mildly changes the demo in `hatch_demo.py` to not be quite so ugly.  The text is probably more important. 

We could merge the demos, but again the problem of losing permanent links is real.  But anyway, that's beyond what I'm interested to do in this PR.  Anyone should feel free to close if they don't deem this an improvement.

File: examples/shapes_and_collections/hatch_demo.py [hunk 1]
```diff
@@ -14,23 +23,26 @@
 y1 = np.arange(1, 5)
 y2 = np.ones(y1.shape) * 4
 
-fig, (ax1, ax2, ax3) = plt.subplots(3)
-
-ax1.bar(x, y1, edgecolor='black', hatch="/")
-ax1.bar(x, y2, bottom=y1, edgecolor='black', hatch='//')
-ax1.set_xticks([1.5, 2.5, 3.5, 4.5])
+fig = plt.figure()
+axs = fig.subplot_mosaic([['bar1', 'patches'], ['bar2', 'patches']])
```
Comment (story645) [inline examples/shapes_and_collections/hatch_demo.py:27]: Happy dance that you're using mosiac (for the record fig, axs = plt.subplot_mosaic also works) 'cause I so think it's well suited for these kinds of things. 
But also not clear why you're reassigning them:
```python
ax = axs['bar1']
ax.bar(x, y2, bottom=y1, edgecolor='black', hatch=['*', 'o', 'O', '.'])
```
versus using 'em directly:
```python
axs['bar'].bar(x, y2, bottom=y1, edgecolor='black', hatch=['*', 'o', 'O', '.'])
```

File: examples/shapes_and_collections/hatch_demo.py [hunk 2]
```diff
@@ -14,23 +23,26 @@
...
+ax = axs['bar2']
+ax.bar(x, y1, edgecolor='black', hatch=['--', '+', 'x', '\\'])
+ax.bar(x, y2, bottom=y1, edgecolor='black', hatch=['*', 'o', 'O', '.'])
 
+ax = axs['patches']
+x = np.arange(0, 40, 0.2)
+ax.fill_between(x, np.sin(x) * 4 + 30, y2=0, hatch='///', zorder=2, fc='c')
+ax.add_patch(Ellipse((4, 50), 10, 10, fill=True, hatch='*', facecolor='y'))
+ax.add_patch(Polygon([(10, 20), (30, 50), (50, 10)], hatch='\\/...',
+                     facecolor='g'))
```
Comment (story645) [inline examples/shapes_and_collections/hatch_demo.py:42]: I'm turning this into a twitter game 'cause 😍

File: examples/shapes_and_collections/hatch_demo.py (hunk 1 above)
Comment (jklymak) [reply examples/shapes_and_collections/hatch_demo.py:27]: Less typing?

Comment (story645) [reply examples/shapes_and_collections/hatch_demo.py:27]: I don't think it's that much less and I think the reassignment especially w/ same variable name is  potential unnecessary source of failure.

Comment (jklymak) [reply examples/shapes_and_collections/hatch_demo.py:27]: ... wrote out the long form

Comment (jklymak) [reply examples/shapes_and_collections/hatch_demo.py:27]: >  I think the reassignment especially w/ same variable name is potential unnecessary source of failure.

I'd argue the opposite - its less prone to typos and easier to refactor.

Comment (story645) [reply examples/shapes_and_collections/hatch_demo.py:27]: I agree it's easier to refactor, but it's prone to leaving out that assignment statement and therefore messing up all the references.

**Pull request merged**